packages/apollo-language-server/src/project/client.ts

```diff
@@ -211,10 +219,69 @@ export class GraphQLClientProject extends GraphQLProject {
   get clientSchema(): DocumentNode {
     return {
       kind: Kind.DOCUMENT,
-      definitions: this.typeSystemDefinitionsAndExtensions
+      definitions: [
+        ...this.typeSystemDefinitionsAndExtensions,
+        ...this.missingApolloClientDirectives
+      ]
     };
   }
 
+  get missingApolloClientDirectives(): readonly DefinitionNode[] {
+    const { serviceSchema } = this;
+
+    const serviceDirectives = serviceSchema
+      ? serviceSchema.getDirectives().map(directive => directive.name)
+      : [];
+
+    const clientDirectives = this.typeSystemDefinitionsAndExtensions
+      .filter(def => def.kind === Kind.DIRECTIVE_DEFINITION)
+      .map(def => (def as DirectiveDefinitionNode).name.value);
```

Comment:
Nothing wrong with this approach, but predicates are a nice way to tell TS that we know the type of something, and it looks like you'd be able to use it in a couple places.

```ts
function isDirectiveDefinitionNode(node: ASTNode): node is DirectiveDefinitionNode {
  return def.kind === Kind.DIRECTIVE_DEFINITION;
}
```

Then `isDirectiveDefinitionNode` becomes your filter function. From there, TS will understand you've got `DirectiveDefinitionNode[]` as a result!

Reply:
This is a great suggestion, but unfortunately it doesn't remove the need to cast the definition to a `DirectiveDefinitionNode `. It seems that TypeScript is unable to modify the return type of the array with the `filter` method:
```javascript
filter(callbackfn: (value: T, index: number, array: ReadonlyArray<T>) => unknown, thisArg?: any): T[];
```
Since the return type is always `T[]`, I'm not sure if there's any way we can avoid casting like this.

Reply:
Nevermind, it turns out you have to do this,
```javascript
const apolloDirectives = apolloAst.definitions
  .filter(isDirectiveDefinitionNode)
  .map(def => def.name.value);
```
NOT this,
```javascript
const apolloDirectives = apolloAst.definitions
  .filter(def => isDirectiveDefinitionNode(def))
  .map(def => def.name.value);
```

Reply:
TIL! That makes sense since the second way is "spoofing" the fn signature in a sense.